### PR TITLE
#12455 - ruby/3.1.0: expose more configuration options for native extension gems

### LINKED ISSUE #12455: [package] ruby/3.1.0: expose more configuration options for native extension gems
### Package and Environment Details

* Package Name/Version: **ruby/3.1.0**
* Operating System+version:  All

### Conan profile

_No response_

### Steps to reproduce

the ruby/3.1.0 recipe does not expose some useful configuration parameters that relate to the native extensions, such as `--with-static-linked-ext` (link external modules statically) and `--enable-load-relative` (resolve load paths at run time).

Additionally, the dependencies for some of these extension gems should probably be provided by conan similarly to what's done for openssl, and should be controlled via the `--with-{ext}-dir=xxx` flags to ensure it is correctly picked up and not the system installed version.

### Logs

_No response_

### PR DESCRIPTION
**Pull request by jmarrec**
Specify library name and version:  **ruby/3.1.0**

- Fix #12455

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/ruby/all/test_package/test_package.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+         puts "I can correctly load one of the extension gems - bigdecimal"
+       rescue Exception => e
+         puts
+         puts "Error: #{e.message}"
+         puts "Backtrace:\n\t" + e.backtrace.join("\n\t")
+         raise
+       end
+     )");
+
+  ruby_finalize();
```
Comment (jmarrec) [inline recipes/ruby/all/test_package/test_package.cpp:54]: Extend testing to require one of the extension gems.

File: recipes/ruby/all/conanfile.py [hunk 2]
```diff
@@ -31,12 +31,26 @@ class RubyConan(ConanFile):
...
         "fPIC": True,
-        "with_openssl": True
+        "with_openssl": True,
+
+        "with_static_linked_ext": True,
+        "with_enable_load_relative": True,
+        "with_libyaml": True,
+        "with_libffi": True,
+        "with_readline": True,
+        'with_gmp': True,
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:53]: New options

File: recipes/ruby/all/conanfile.py [hunk 3]
```diff
@@ -61,10 +75,22 @@ def _msvc_optflag(self):
...
+            self.requires("libyaml/0.2.5")
+
+        if self.options.with_libffi:
+            self.requires("libffi/3.4.2")
+
+        if self.options.with_readline:
+            self.requires("readline/8.1.2")
+
+        if self.options.with_gmp:
+            self.requires("gmp/6.2.1")
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:92]: Control dependencies

File: recipes/ruby/all/conanfile.py [hunk 4]
```diff
@@ -77,8 +103,12 @@ def validate(self):
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
+            del self.options.with_static_linked_ext
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:106]: This only makes sense in the case of a static build to begin with

File: recipes/ruby/all/conanfile.py [hunk 5]
```diff
@@ -77,8 +103,12 @@ def validate(self):
...
         del self.settings.compiler.libcxx
         del self.settings.compiler.cppstd
+        if self.settings.os == 'Windows':
+            # readline isn't supported on Windows
+            self.options.with_readline = False
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:111]: readline caveat on windows.

File: recipes/ruby/all/conanfile.py [hunk 6]
```diff
@@ -107,6 +137,17 @@ def generate(self):
             if "--enable-shared" not in tc.configure_args:
                 tc.configure_args.append("--enable-shared")
 
+        if not self.options.shared and self.options.with_static_linked_ext:
+            tc.configure_args.append('--with-static-linked-ext')
+
+        if self.options.with_enable_load_relative:
+            tc.configure_args.append('--enable-load-relative')
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:144]: Pass these two to ruby configure.

File: recipes/ruby/all/conanfile.py [hunk 7]
```diff
@@ -107,6 +137,17 @@ def generate(self):
...
+
+        for name, dep_cpp_info in self.deps_cpp_info.dependencies:
+            if name in ['zlib', 'openssl', 'libffi', 'libyaml', 'readline', 'gmp']:
+                root_path = tools.unix_path(dep_cpp_info.rootpath)
+                tc.configure_args.append(f'--with-{name}-dir={root_path}')
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:149]: For dependencies, pass the --with-ext-dir flag to ensure ruby find the one from conan and not the system installed one.

File: recipes/ruby/all/conanfile.py [hunk 8]
```diff
@@ -155,6 +196,13 @@ def package(self):
         tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
         tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.pdb")
 
+        # install the enc/*.a / ext/*.a libraries
+        if not self.options.shared and self.options.with_static_linked_ext:
+            for dirname in ['ext', 'enc']:
+                dst = os.path.join('lib', dirname)
+                self.copy('*.a', dst=dst, src=dirname, keep_path=True)
+                self.copy('*.lib', dst=dst, src=dirname, keep_path=True)
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:204]: Ship the potential static linked exts too.

File: recipes/ruby/all/conanfile.py [hunk 9]
```diff
@@ -173,9 +221,19 @@ def package_info(self):
...
+        if self.options.with_gmp:
+            rubylib.requires.append("gmp::gmp")
         if self.options.with_openssl:
             rubylib.requires.append("openssl::openssl")
+        if self.options.with_libyaml:
+            rubylib.requires.append("libyaml::libyaml")
+        if self.options.with_libffi:
+            rubylib.requires.append("libffi::libffi")
+        if self.options.with_readline:
+            rubylib.requires.append("readline::readline")
```
Comment (jmarrec) [inline recipes/ruby/all/conanfile.py:235]: adjust the requires. I'm not sure about that TODO...

**Comment by jmarrec** [comment]
Fun two builds are throwing a segfault on linux...

```
compiling parser.c
compiling parserstate.c
during GIMPLE pass: unswitch
parserstate.c: In function â€˜parser_typevar_memberâ€™:
parserstate.c:73:6: internal compiler error: Segmentation fault
   73 | bool parser_typevar_member(parserstate *state, ID id) {
      |      ^~~~~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-9/README.Bugs> for instructions.
```

**Comment by uilianries** [comment]
@jmarrec Thank you for your contribution, really good to see all these new options. Please, take a look on v2 Linter result. Some conans.tools are now deprecated and should move toward to Conan v2 approach: https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md#import-tools-from-conan

**Comment by jmarrec** [comment]
@uilianries I had ported locally to v2 but then I saw Eric's bot point to #12208 which already does it so I didn't push it, assuming I should wait for that one to merge first?

**Comment by uilianries** [comment]
@jmarrec If it works, please, go ahead, we can merge your PR first. I can adapt my PR later.

**Comment by jmarrec** [comment]
I'm actively trying to understand why the build_shared=True is failing and build_shared=False is not on gcc (eg 8, but I can reproduce on conanio/gcc10:latest too)

When build_shared=False, --with-static-linked-ext is passed, and therefore you see in the log:
```
skip installation of rbs 2.0.0; bundled gem with an extension library is not supported on --with-static-linked-ext
```

When build_shared=True, it tries to build https://github.com/ruby/rbs/ (at version 2.1.0) but it gets a GCC ICE (internal compilation error):

```
/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/lib/rubygems/ext/builder.rb:95:in `run': ERROR: Failed to build gem native extension. (Gem::Ext::BuildError)

    current directory: /home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/bin/ruby --disable\\=gems -I/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d//lib/ruby/3.1.0/x86_64-linux -I /home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/lib -r ./siteconf20220914-130384-gcxbur.rb extconf.rb
creating Makefile

current directory: /home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
make DESTDIR\\= clean
make[1]: Entering directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'
make[1]: Leaving directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'

current directory: /home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
make DESTDIR\\=
make[1]: Entering directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'
compiling constants.c
compiling lexer.c
compiling lexstate.c
compiling main.c
compiling location.c
compiling parser.c
compiling parserstate.c
compiling ruby_objs.c
compiling unescape.c
during GIMPLE pass: unswitch
parserstate.c: In function ‘parser_typevar_member’:
parserstate.c:73:6: internal compiler error: Segmentation fault
 bool parser_typevar_member(parserstate *state, ID id) {
      ^~~~~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
Makefile:245: recipe for target 'parserstate.o' failed
make[1]: *** [parserstate.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'

make failed, exit code 2
```

Why it does that... I'm not sure at all

**Comment by jmarrec** [comment]
Inside the conanio/gcc8, I launch `conan create . ruby/3.1.0@ -b missing -o ruby:shared=True` and let it fail.

then I cd into the build dir (the sha will vary), and recreate the command that is being launched

```
cd ~/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release
./miniruby -I/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/lib -I. -I.ext/common  /home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/tool/runruby.rb --extout=.ext  -- --disable-gems -r./x86_64-linux-fake /home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/tool/rbinstall.rb --make="make" --dest-dir="/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d" --extout=".ext" --ext-build-dir="./ext" --mflags="-j --jobserver-fds=4,5" --make-flags="--verbose -j --jobserver-fds=4,5 -- DESTDIR=/home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d" --data-mode=0644 --prog-mode=0755 --installed-list .installed.list --mantype="doc"  --gnumake --install=all --exclude=do
```

It fails the same way:

```
during GIMPLE pass: unswitch
parserstate.c: In function 'parser_typevar_member':
parserstate.c:73:6: internal compiler error: Segmentation fault
 bool parser_typevar_member(parserstate *state, ID id) {
      ^~~~~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
Makefile:245: recipe for target 'parserstate.o' failed
make: *** [parserstate.o] Error 1
```

I modified `/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/tool/runruby.rb` to add a `puts env` to see the variables being set

```
{"RUBY_THREAD_MACHINE_STACK_SIZE"=>"1",
 "RUBY_FIBER_MACHINE_STACK_SIZE"=>"1",
 "RUBY"=>"/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/ruby",
 "GEM_HOME"=>"/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/.bundle",
 "GEM_PATH"=>"/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/.bundle",
 "GEM_COMMAND"=>
  "/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/ruby -rrubygems /home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/bin/gem --backtrace",
 "PATH"=>
  "/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release:/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release:/opt/pyenv/plugins/pyenv-virtualenv/shims:/opt/pyenv/shims:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
 "RUBYLIB"=>
  "/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release:/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/.ext/common:/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/.ext/x86_64-linux:/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/src/lib",
 "LD_LIBRARY_PATH"=>"/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release",
 "LD_PRELOAD"=>"/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/libruby.so.3.1.0"
```

----

If I go to the rbs_extension directory and `make`, it works just fine:

```
cd /home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
make
=> Success
```

Now If I define the LD_PRELOAD, it fails with the segfault

```
cd /home/conan/.conan/data/ruby/3.1.0/_/_/package/ec66712f646013b4707f8fffdde5490aa3732a9d/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
export LD_PRELOAD="/home/conan/.conan/data/ruby/3.1.0/_/_/build/ec66712f646013b4707f8fffdde5490aa3732a9d/build-release/libruby.so.3.1.0"
make
=> parserstate.c:73:6: internal compiler error: Segmentation fault
```

**Pull request closed without merging**